### ISSUE TITLE
False IgnoredReturnValue positive

### ISSUE THREAD

**Issue by ahibrahimleague** [detekt/detekt#6134]
## Expected Behavior
This block of code flags a detekt violation of `IgnoredReturnValue`

Meaningless block of code for demonstration purposes:

```
internal class MyUseCase {
    fun execute(): Flow<Int> = flow {
        emit(1)
    }
}

internal class MyUseCaseTest {
    @Test
    fun `test myusecase`() = runTest {
        val mockUseCase = mockk<MyUseCase>()

        verify { mockUseCase.execute() }
    }
}
```

## Observed Behavior
Using detekt 1.22, when running Detekt with TR (`./gradlew detektTest`), Detekt task fails with the below error

> The call execute is returning a value that is ignored. [IgnoredReturnValue]

## Steps to Reproduce
Just paste the above block of code in one of the test files, and run `./gradlew detektTest`.

## Your Environment

* Version of detekt used: 1.22
* Version of Gradle used (if applicable): 7.6
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version: Macbook Apple M1 Pro Ventura 13.3.1
* Link to your project (if it's a public repository): Not public.

**Comment by ahibrahimleague** [comment]
@cortinico I think that's because it's of type Flow, and Coroutine Flow shouldn't be ignored. 
https://github.com/detekt/detekt/blob/dcfb8415c3ae9864d9e25226faebd26713bdab90/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/IgnoredReturnValue.kt#L88-L93

**Comment by TWiStErRob** [comment]
Looks like this would help
```yml
IgnoredReturnValue:
  ignoreFunctionCall:
    - com.example.MyUseCase.execute
```
but it doesn't scale.

**Comment by TWiStErRob** [comment]
What's the configuration for this rule in your project?

I tried adding to IgnoredReturnValueSpec in detekt, but it passed on both v1.22.0 and main (v1.23.0+) tags:
```kotlin
    @Test
    fun `repro6134`() {
        val code = """
            internal class MyUseCase {
                fun execute(): Flow<Int> = flow {
                    emit(1)
                }
            }
            internal class MyUseCaseTest {
                @Test
                fun `test myusecase`() = runTest {
                    val mockUseCase = mockk<MyUseCase>()
            
                    verify { mockUseCase.execute() }
                }
            }
            annotation class Test
            fun runTest(block: suspend () -> Unit): Unit = TODO()
            fun <T> mockk(): T = TODO()
            fun verify(block: () -> Unit): Unit = TODO()
            interface Flow<T>
            fun <T> flow(block: suspend () -> Unit): Flow<T> = TODO()
            suspend fun emit(value: Any): Unit = TODO()
        """.trimIndent()
        val findings = IgnoredReturnValue().compileAndLintWithContext(env, code)
        assertThat(findings).isEmpty()
    }
```

**Comment by cortinico** [comment]
> @cortinico I think that's because it's of type Flow, and Coroutine Flow shouldn't be ignored.

Yup you're right as the conditions are in OR and not in AND

**Comment by ahibrahimleague** [comment]
@TWiStErRob Here's a reproducer project [DetektIgnoreReturnValueRepro.zip](https://github.com/detekt/detekt/files/11574523/DetektIgnoreReturnValueRepro.zip), run `./gradlew detektTest`

and you'll hit by 

> MyUseCaseTest.kt:20:30: The call execute is returning a value that is ignored. [IgnoredReturnValue]

I think in your test code above, it didn't catch the issue because the Flow stub is not defined in the correct package (instead of `kotlinx.coroutines.flow.Flow`), so the rule didn't capture the return value as part of the condition.

**Comment by TWiStErRob** [comment]
Ah, good point about the package, thanks. Replaced whitelist with `*Flow` locally and it correctly fails now.

What can we do here? The `IgnoredReturnValue` logic is quite complex:
 * Check all the annotated functions
 * Or those who have a specific return types
 * But ignore some functions by signature
 * This issue would add: Except in some call contexts

I think you can configure detekt here with an [annotation suppressor](https://detekt.dev/docs/introduction/suppressors/#annotation-suppressor) including the `org.junit.Test` annotation. This is not as ideal as using a (non-existent) enclosing lambda suppressor for `mockk.verify { }`, but it sounds good enough.

**Comment by BraisGabin** [comment]
Agree. I think this is not a false positive. The return value *is* ignored. So we should find a way to suppress this cases instead of changing the rule.

**Comment by TWiStErRob** [comment]
Actually, @BraisGabin triggered my brain, the root cause here is the signature of `verify`:
https://github.com/mockk/mockk/blob/master/modules/mockk/src/commonMain/kotlin/io/mockk/MockK.kt#LL209C1-L217C2
in particular: `verifyBlock: MockKVerificationScope.() -> Unit`
If that was `verifyBlock: MockKVerificationScope.() -> Any?`, there would be no problem, right? So another "fix" is to duplicate the signature of `verify` as an inline function and delegate to the original one.

**Comment by schalkms** [comment]
How do we move forward with this issue? We have a separate one, namely #5510.
Shall we close this issue?

**Comment by BraisGabin** [comment]
Ya, I think that we can close this one.
